sqlmesh/core/snapshot/evaluator.py

```diff
@@ -1786,9 +1795,30 @@ def migrate(
         snapshot: Snapshot,
         **kwargs: t.Any,
     ) -> None:
-        # Not entirely true, many engines support modifying some of the metadata fields on a managed table
-        # eg Snowflake allows you to ALTER DYNAMIC TABLE foo SET WAREHOUSE=my_other_wh;
-        raise ConfigError(f"Cannot mutate managed table: {target_table_name}")
+        potential_alter_expressions = self.adapter.get_alter_expressions(
+            target_table_name, source_table_name
+        )
+        if len(potential_alter_expressions) > 0:
+            # this can happen when a forward-only change is applied directly to prod without going via a dev environment first
+            # in the normal case, the current prod snapshot is mutated via ALTER statements to match the schema of the dev preview table
+            # However, since we cant change the schema of a managed model, we have to create a new one instead
+            logger.info(
+                "Creating new managed table: %s to migrate %s to the schema of %s",
+                target_table_name,
+                snapshot.model.name,
+                source_table_name,
+            )
+
+            self.create(
```

Comment:
Can't dynamic tables be altered? I'm concerned that with this approach all prior data will be wiped out and the table will have to start from scratch which defeats the purpose of a forward-only change.

Reply:
Yes, that's true. A dynamic table is essentially a materialized view where the database keeps it up to date internally, so you cant `ALTER TABLE ADD COLUMN` to it, to change the schema you have to replace the query backing the view which recreates it.

Would it be better to keep the original behavior of just throwing an error (although only do it if there are schema changes to be made) and declaring that forward-only changes aren't supported on managed models?

Reply:
by the way - a side-effect of recreating managed models declared as forward-only is that this provides a pattern for users to update managed models without having a bunch of old snapshot versions lying around (that can incur compute cost).

I believe Nick was [asking about this recently](https://tobiko-data.slack.com/archives/C044BRE5W4S/p1721419318871229) and resorted to writing his own cleanup logic

Reply:
> by the way - a side-effect of recreating managed models declared as forward-only is that this provides a pattern for users to update 

I can see that. However, this goes against the very idea of forward-only which suppose to prevent the **rebuild**. That's why I like the 

> Would it be better to keep the original behavior of just throwing an error (although only do it if there are schema changes to be made) and declaring that forward-only changes aren't supported on managed models?

suggestion better.

Reply:
Ok, if there are no schema changes then `migrate()` is a no-op which should resolve the issue of an unrelated macro change tagging the model as modified and hitting `migrate()`  codepath.

If there are schema changes, an error is thrown advising the user that managed models cannot be forward only